library/core/src/pin.rs

```diff
@@ -802,6 +802,44 @@ impl<T: ?Sized> Pin<&'static T> {
     }
 }
 
+impl<'a, P: DerefMut> Pin<&'a mut Pin<P>> {
+    /// Gets a pinned mutable reference from this nested pinned pointer.
+    ///
+    /// This is a generic method to go from `Pin<&mut Pin<Pointer<T>>>` to `Pin<&mut T>`. It is
+    /// safe because the existence of a `Pin<Pointer<T>>` ensures that the pointee, `T`, cannot
+    /// move in the future, and this method does not enable the pointee to move. "Malicious"
+    /// implementations of `Pointer::DerefMut` are likewise ruled out by the contract of
+    /// `Pin::new_unchecked`.
+    #[unstable(feature = "pin_deref_mut", issue = "none")]
+    #[inline(always)]
+    pub fn as_deref_mut(self) -> Pin<&'a mut P::Target> {
+        // SAFETY: What we're asserting here is that going from
+        //
+        //     Pin<&mut Pin<P>>
+        //
+        // to
+        //
+        //     Pin<&mut P::Target>
+        //
+        // is safe.
```

Comment:
My own justification attempts to follow the invariant I sketched [here](https://internals.rust-lang.org/t/unsoundness-in-pin/11311/113).

First of all, it seems like `Pin` might actually be `Unpin`. This might sound surprising, but really what `Unpin` means is that the "pinned" and "not-pinned" typestates of a type are the same, and `Pin` is *always* referring to pinned data, so whether it is in one or the other typestate makes no difference.

Thus we can turn `Pin<&'a mut Pin<P>>` into `&'a mut Pin<P>` via `get_mut`. Now we just call `as_mut` and we are done.